File: src/main/java/no/fintlabs/security/TokenParsingUtils.java
package no.fintlabs.security;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import no.fintlabs.model.User;
import org.springframework.security.core.Authentication;
import org.springframework.security.oauth2.server.resource.authentication.JwtAuthenticationToken;
import org.springframework.stereotype.Service;
import org.springframework.util.StringUtils;

import java.util.Map;
import java.util.Optional;
import java.util.UUID;

import static org.springframework.util.StringUtils.hasText;

@Service
@RequiredArgsConstructor
@Slf4j
public class TokenParsingUtils {

    public AuditorScope bindAuditor(Authentication authentication) {
        tryGetUser(authentication)
                .map(u -> firstNonBlank(toTitleCase(u.getName()), u.getEmail()))
                .filter(StringUtils::hasText)
                .ifPresent(CurrentAuditor::set);

        return CurrentAuditor::clear;
    }

    private static String toTitleCase(String input) {
        if (!hasText(input)) return input;

        StringBuilder result = new StringBuilder(input.length());
        boolean startOfWord = true;

        for (char c : input.trim().toCharArray()) {
            if (Character.isLetter(c)) {
                result.append(startOfWord ? Character.toTitleCase(c) : Character.toLowerCase(c));
                startOfWord = false;
            } else {
                result.append(c);
                startOfWord = Character.isWhitespace(c) || c == '-' || c == '\'' || c == '’';
            }
        }
        return result.toString();
    }

    public Optional<User> tryGetUser(Authentication authentication) {
        if (!(authentication instanceof JwtAuthenticationToken jwt)) return Optional.empty();

        Map<String, Object> attrs = jwt.getTokenAttributes();

        String name = firstNonBlank(
                text(attrs.get("displayname")),
                text(attrs.get("name")),
                text(attrs.get("preferred_username")),
                text(attrs.get("principalName"))
        );

        String email = firstNonBlank(
                text(attrs.get("email")),
                text(attrs.get("principalName"))
        );

        UUID oid = parseUuid(text(attrs.get("objectidentifier")));

        if ((name == null && email == null) || oid == null) return Optional.empty();

        return Optional.of(User.builder()
                .objectIdentifier(oid)
                .name(name)
                .email(email)
                .build());
    }

    private static String text(Object o) {
        return Optional.ofNullable(o)
                .map(Object::toString)
                .map(String::strip)
                .filter(StringUtils::hasText)
                .orElse(null);
    }

    private static String firstNonBlank(String... values) {
        if (values == null) return null;
        for (String v : values) {
            if (hasText(v)) return v;
        }
        return null;
    }

    private static UUID parseUuid(String s) {
        try {
            return s == null ? null : UUID.fromString(s);
        } catch (IllegalArgumentException e) {
            return null;
        }
    }
}

File: src/main/kotlin/no/novari/flyt/configuration/security/TokenParsingUtils.kt
package no.novari.flyt.configuration.security

import no.novari.flyt.configuration.model.User
import org.springframework.security.core.Authentication
import org.springframework.security.oauth2.server.resource.authentication.JwtAuthenticationToken
import org.springframework.stereotype.Service
import java.util.UUID

@Service
class TokenParsingUtils {
    fun bindAuditor(authentication: Authentication): AuditorScope {
        tryGetAuditorName(authentication)
            ?.takeUnless(String::isBlank)
            ?.let(CurrentAuditor::set)

        return AuditorScope { CurrentAuditor.clear() }
    }

    fun tryGetAuditorName(authentication: Authentication): String? {
        if (authentication !is JwtAuthenticationToken) {
            return null
        }

        val attrs = authentication.tokenAttributes
        val displayName =
            firstNonBlank(
                text(attrs["displayname"]),
                text(attrs["name"]),
            )
        val email =
            firstNonBlank(
                text(attrs["email"]),
                text(attrs["preferred_username"]),
                text(attrs["principalName"]),
            )

        return firstNonBlank(toTitleCase(displayName), email)
    }

    fun tryGetUser(authentication: Authentication): User? {
        if (authentication !is JwtAuthenticationToken) {
            return null
        }

        val attrs = authentication.tokenAttributes

        val name =
            firstNonBlank(
                text(attrs["displayname"]),
                text(attrs["name"]),
                text(attrs["preferred_username"]),
                text(attrs["principalName"]),
            )

        val email =
            firstNonBlank(
                text(attrs["email"]),
                text(attrs["principalName"]),
            )

        val oid = parseUuid(text(attrs["objectidentifier"]))

        if ((name == null && email == null) || oid == null) {
            return null
        }

        return User
            .builder()
            .objectIdentifier(oid)
            .name(name)
            .email(email)
            .build()
    }

    private fun text(value: Any?): String? = value?.toString()?.trim()?.takeUnless(String::isBlank)

    private fun firstNonBlank(vararg values: String?): String? = values.firstOrNull { !it.isNullOrBlank() }

    private fun parseUuid(value: String?): UUID? =
        try {
            value?.let(UUID::fromString)
        } catch (_: IllegalArgumentException) {
            null
        }

    private companion object {
        @JvmStatic
        private fun toTitleCase(input: String?): String? {
            val trimmed = input?.trim().takeUnless(String?::isNullOrBlank) ?: return input
            val result = StringBuilder(trimmed.length)
            var startOfWord = true

            for (char in trimmed) {
                if (char.isLetter()) {
                    result.append(if (startOfWord) char.titlecaseChar() else char.lowercaseChar())
                    startOfWord = false
                } else {
                    result.append(char)
                    startOfWord = char.isWhitespace() || char == '-' || char == '\'' || char == '’'
                }
            }

            return result.toString()
        }
    }
}
